src/Build/Evaluation/LazyItemEvaluator.cs

```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
+                        // This is a rare operation, don't bother updating the dictionary for now. It will be recreated as needed.
+                        _dictionaryBuilder = null;
+                    }
+                }
+
+                public void Add(ItemData data)
+                {
+                    _listBuilder.Add(data);
+                    if (_dictionaryBuilder != null)
+                    {
+                        AddToDictionary(data.Item);
+                    }
+                }
+
+                public void Clear()
+                {
+                    _listBuilder.Clear();
+                    _dictionaryBuilder?.Clear();
+                }
+
+                public void RemoveAll(ICollection<I> itemsToRemove)
+                {
+                    _listBuilder.RemoveAll(item => itemsToRemove.Contains(item.Item));
+                    // This is a rare operation, don't bother updating the dictionary for now. It will be recreated as needed.
+                    _dictionaryBuilder = null;
+                }
+
+                /// <summary>
+                /// Removes items from the collection that match the given ItemSpec.
+                /// </summary>
+                /// <remarks>
+                /// If <see cref="_dictionaryBuilder"/> does not exist yet, it is created in this method to avoid the cost of comparing each item
+                /// being removed with each item already in the collection. The dictionary is kept in sync with the <see cref="_listBuilder"/>
+                /// as long as practical. If an operation would result in too much of such work, the dictionary is simply dropped and recreated
+                /// later if/when needed.
+                /// </remarks>
+                public void RemoveMatchingItems(ItemSpec<P, I> itemSpec)
+                {
+                    HashSet<I> items = null;
+                    var dictionaryBuilder = GetOrCreateDictionaryBuilder();
+                    foreach (var fragment in itemSpec.Fragments)
+                    {
+                        IEnumerable<string> referencedItems = fragment.GetReferencedItems();
+                        if (referencedItems != null)
+                        {
+                            // The fragment can enumerate its referenced items, we can do dictionary lookups.
+                            foreach (var spec in referencedItems)
+                            {
+                                string key = FileUtilities.NormalizePathForComparisonNoThrow(spec, fragment.ProjectDirectory);
+                                if (dictionaryBuilder.TryGetValue(key, out var multiValue))
+                                {
+                                    items ??= new HashSet<I>();
+                                    foreach (I item in multiValue)
+                                    {
+                                        items.Add(item);
+                                    }
```

Comment:
```suggestion
                                    items.UnionWith(multiValue);
```

Reply:
Done, thank you!

Reply:
I reverted this because it doesn't work well with the struct enumerator of `DictionaryValue`. A manual `foreach` avoids boxing and the expected number of items added is 1 so `UnionWith` wouldn't be helping even if it was optimized for multiple adds (which it looks like it is not).